### PR TITLE
Engine Brotli compressed and optional http servers compression

### PR DESCRIPTION
**Pull request by JT117**
Relates to #199 

### Goal 1: Serve brotli pre compressed perseus engine files 

- Change perseus-cli to brotli compress the engine files after building them (maybe better to add it in the cmds vec ?)
- Add a flag in the perseus-cli to deactivate the brotli compress file generation: disable_engine_compression
- Change the servers configuration to serve the .br files first and if not found defaults to the uncompressed file.

#### Result :
On the basic example: 
| Mode            | File                | raw kB      |  brotli kB  |
| ------------- | -------------  | ---------- | ---------- |
| serve            | bundle.js        |  42.42       | 6.5           |
|                     | bundle.wasm  | 2980        | 554.71     |
| deploy          | bundle.js        |  32.43       | 5.23         |
|                     | bundle.wasm  | 315.98      | 109.62     |

The lighthouse score on mobile goes up from 97% to 100%
On my app it goes from 89% to 100% on mobile.

Remarks: Each server has it's own way to do thing, it goes from simple to I had to make a new Responder for Rocket

### Goal 2: Add the possibility for users to enable http compression on the servers

- Add on actix-web, axum, rocket and warp a new features: dflt-server-with-compression
- If the features is enabled, the server activate native compression

#### Result:
On the basic example: 

|     Server | File     | raw (kB) | compressed (kB) | Type   |
|----------|-------|----------|------------------| -------|
| rocket      | index    | 1.010           | 0.638  | brotli |
|                | about    | 0.927           | 0.328  | brotli |
| warp        | index    | 1.010           | 0.657  | gzip   |
|                 | about    | 0.927           | 0.611  | gzip   |
| actix-web | index    | 1.010           | 0.694  | gzip   |
|                 | about    | 0.927           | 0.647  | gzip   |
| axum       | index    | 1.010           | 0.657  | gzip   |
|                | about    | 0.927           | 0.610  | gzip   |

I would be happy to discuss the changes 😄 , and explain the choices that I made

### CODE REVIEW

File: packages/perseus-actix-web/src/lib.rs [hunk 1]
```diff
@@ -160,15 +161,43 @@ pub async fn configurer<M: MutableStore + 'static, T: TranslationsManager + 'sta
...
+async fn wasm_js_bundle(opts: web::Data<ServerOptions>) -> std::io::Result<CustomizeResponder<NamedFile>> {
+    search_for_pre_compressed_version(&opts.wasm_js_bundle, "application/javascript; charset=utf-8".to_string())
 }
-async fn wasm_js_bundle(opts: web::Data<ServerOptions>) -> std::io::Result<NamedFile> {
-    NamedFile::open(&opts.wasm_js_bundle)
+
+fn search_for_pre_compressed_version(path: &str, application_type: String) -> std::io::Result<CustomizeResponder<NamedFile>> {
+    let pre_compressed_path = format!("{}.br", path);
+    match NamedFile::open(pre_compressed_path) {
+        Ok(file) => Ok( file.customize()
```
Comment (arctic-hen7) [inline packages/perseus-actix-web/src/lib.rs:179]:
```suggestion
        Ok(file) => Ok(file.customize()
```

File: packages/perseus-actix-web/src/lib.rs [hunk 2]
```diff
@@ -160,15 +161,43 @@ pub async fn configurer<M: MutableStore + 'static, T: TranslationsManager + 'sta
...
+                "Content-Encoding".to_string(),
+                "br".to_string(),
+            ))
+            .insert_header((
+                "Content-Type".to_string(),
+                application_type,
+            ))),
+        Err(_) => {
+            match NamedFile::open(path) {
+                Ok(file) => Ok( file.customize()
```
Comment (arctic-hen7) [inline packages/perseus-actix-web/src/lib.rs:190]:
```suggestion
                Ok(file) => Ok(file.customize()
```

File: packages/perseus-actix-web/src/lib.rs [hunk 3]
```diff
@@ -214,3 +243,36 @@ pub async fn dflt_server<M: MutableStore + 'static, T: TranslationsManager + 'st
...
         .expect("Server failed.") // TODO Improve error message here
 }
+
+
+/// Creates and starts the default Perseus server with GZIP compression enabled using Actix Web. This should
+/// be run in a `main()` function annotated with `#[tokio::main]` (which
+/// requires the `macros` and `rt-multi-thread` features on the `tokio`
+/// dependency).
+#[cfg(feature = "dflt-server-with-compression")]
+pub async fn dflt_server<M: MutableStore + 'static, T: TranslationsManager + 'static>(
```
Comment (arctic-hen7) [inline packages/perseus-actix-web/src/lib.rs:253]:
```suggestion
pub async fn dflt_server_with_compression<M: MutableStore + 'static, T: TranslationsManager + 'static>(
```

File: packages/perseus-axum/Cargo.toml [hunk 4]
```diff
@@ -22,6 +22,7 @@ tower-http = { version = "0.3", features = [ "fs" ] }
 [features]
 # Enables the default server configuration, which provides a convenience function if you're not adding any extra routes
 dflt-server = []
+dflt-server-with-compression = ["tower-http/compression-gzip"]
```
Comment (arctic-hen7) [inline packages/perseus-axum/Cargo.toml:25]:
```suggestion
dflt-server-with-compression = [ "tower-http/compression-gzip" ]
```

File: packages/perseus-axum/src/lib.rs [hunk 5]
```diff
@@ -185,7 +192,34 @@ pub async fn dflt_server<M: MutableStore + 'static, T: TranslationsManager + 'st
...
+        .serve(app.into_make_service())
+        .await
+        .unwrap();
+}
+
+/// Creates and starts the default Perseus server with compression using Axum. This should be run
+/// in a `main` function annotated with `#[tokio::main]` (which requires the
+/// `macros` and `rt-multi-thread` features on the `tokio` dependency).
+#[cfg(feature = "dflt-server-with-compression")]
+pub async fn dflt_server<M: MutableStore + 'static, T: TranslationsManager + 'static>(
```
Comment (arctic-hen7) [inline packages/perseus-axum/src/lib.rs:208]:
```suggestion
pub async fn dflt_server_with_compression<M: MutableStore + 'static, T: TranslationsManager + 'static>(
```

File: packages/perseus-cli/src/build.rs [hunk 6]
```diff
@@ -152,6 +157,29 @@ pub fn build_internal(
                 verbose,
             )?);
 
+            // If the compression is enable (enabled by default), we brotli compress each file in the `dist/pkg` directory
```
Comment (arctic-hen7) [inline packages/perseus-cli/src/build.rs:160]:
```suggestion
            // If compression is enabled (which it is by default), we compress each file with Brotli in `dist/pkg`
```

File: packages/perseus-cli/src/build.rs [hunk 7]
```diff
@@ -152,6 +157,29 @@ pub fn build_internal(
...
+            // Make this a step in the build process with a dedicated spinner ? How to make it work in the cmds arrays ?
```
Comment (arctic-hen7) [inline packages/perseus-cli/src/build.rs:161]:
```suggestion
```

File: packages/perseus-cli/src/build.rs
Comment (arctic-hen7) [inline packages/perseus-cli/src/build.rs:1]: There are only ever two files in `dist/pkg/` that matter: the Wasm bundle and the JS bundle. The other two are types files that are presently unused, and compressing them seems unnecessary. I would suggest simplifying this code by avoiding walking the directory and instead just directly compressing those two files.

File: packages/perseus-cli/src/parse.rs [hunk 8]
```diff
@@ -88,6 +88,10 @@ pub struct Opts {
     /// commands are more useful for targeted debugging
     #[clap(long, global = true)]
     pub verbose: bool,
+
+    /// Deactivate the brotli compression of the engine files
```
Comment (arctic-hen7) [inline packages/perseus-cli/src/parse.rs:92]:
```suggestion
    /// Disable Brotli compression of JS and Wasm bundles (may degrade performance)
```

File: packages/perseus-rocket/Cargo.toml [hunk 9]
```diff
@@ -14,10 +14,12 @@ categories = ["wasm", "web-programming::http-server", "development-tools", "asyn
 [dependencies]
 perseus = { path = "../perseus", version = "0.4.2"}
 rocket = "0.5.0-rc.2"
+rocket_async_compression_lib = { version = "0.5.0", optional = true, package = "rocket_async_compression" }
```
Comment (arctic-hen7) [inline packages/perseus-rocket/Cargo.toml:17]: Why not use the regular package name?

File: packages/perseus-rocket/Cargo.toml [hunk 10]
```diff
@@ -14,10 +14,12 @@ categories = ["wasm", "web-programming::http-server", "development-tools", "asyn
...
 
 [features]
 # Enables the default server configuration, which provides a convenience function if you're not adding any extra routes
 dflt-server = []
+dflt-server-with-compression = ["rocket_async_compression_lib"]
```
Comment (arctic-hen7) [inline packages/perseus-rocket/Cargo.toml:22]:
```suggestion
dflt-server-with-compression = [ "rocket_async_compression_lib" ]
```

File: packages/perseus-rocket/src/lib.rs [hunk 11]
```diff
@@ -376,9 +378,36 @@ pub async fn dflt_server<M: MutableStore + 'static, T: TranslationsManager + 'st
...
     if let Err(err) = app.launch().await {
         eprintln!("Error lauching Rocket app: {}.", err);
     }
 }
+
+/// Creates and starts the default Perseus server with compression using Rocket. This should be
+/// run in a `main` function annotated with `#[tokio::main]` (which requires the
+/// `macros` and `rt-multi-thread` features on the `tokio` dependency).
+#[cfg(feature = "dflt-server-with-compression")]
+pub async fn dflt_server<M: MutableStore + 'static, T: TranslationsManager + 'static>(
```
Comment (arctic-hen7) [inline packages/perseus-rocket/src/lib.rs:393]:
```suggestion
pub async fn dflt_server_with_compression<M: MutableStore + 'static, T: TranslationsManager + 'static>(
```

File: packages/perseus-rocket/src/pre_compressed_named_file.rs
Comment (arctic-hen7) [inline packages/perseus-rocket/src/pre_compressed_named_file.rs:1]: As per the whole-review comment, is all this really necessary?

File: packages/perseus-warp/Cargo.toml [hunk 12]
```diff
@@ -20,6 +20,7 @@ warp = { package = "warp-fix-171", version = "0.3" } # Temporary until Warp #171
 [features]
 # Enables the default server configuration, which provides a convenience function if you're not adding any extra routes
 dflt-server = []
+dflt-server-with-compression = ["warp/compression"]
```
Comment (arctic-hen7) [inline packages/perseus-warp/Cargo.toml:23]:
```suggestion
dflt-server-with-compression = [ "warp/compression" ]
```

File: packages/perseus-warp/src/lib.rs [hunk 13]
```diff
@@ -205,3 +239,21 @@ pub async fn dflt_server<M: MutableStore + 'static, T: TranslationsManager + 'st
     let routes = perseus_routes(turbine, opts).await;
     warp::serve(routes).run(addr).await;
 }
+
+/// Creates and starts the Warp Perseus server with compression enable. This should be run
+/// in a `main` function annotated with `#[tokio::main]` (which requires the
+/// `macros` and `rt-multi-thread` features on the `tokio` dependency).
+#[cfg(feature = "dflt-server-with-compression")]
+pub async fn dflt_server<M: MutableStore + 'static, T: TranslationsManager + 'static>(
```
Comment (arctic-hen7) [inline packages/perseus-warp/src/lib.rs:247]:
```suggestion
pub async fn dflt_server_with_compression<M: MutableStore + 'static, T: TranslationsManager + 'static>(
```

**Review by arctic-hen7 — CHANGES_REQUESTED** [review]
Thank you so much for all this, it's really impressive! The benchmarks speak for themselves I think, and this is definitely something I want to get into Perseus sooner rather than later: my only big question is about the Rocket implementation, because it seems like a lot of code for something that's fairly simple in the other servers. Could you explain why all this is necessary? (I've tried to keep the server integrations as simple as possible, is there perhaps some way you could think of refactoring the core so we minimise the amount of code for the servers?)

As for having compression in the CLI have its own spinner, how long have you observed compression taking for the examples? Also, I would suggest having the `dflt-server-with-compression` flag activate a function with a unique name so the features aren't mutually exclusive (who knows, there may be some exotic configuration using both, but I don't see any reason to force the user into only having one). Also, I recommend changing `disable_engine_compression` to `disable_bundle_compression` to make what that flag does clearer by name in the CLI. I've left some more detailed minor suggestions (whitespace etc.) in my review.

Also, do you want to move the compression process in the CLI to the deployment system? That would avoid unnecessarily slowing down development builds (which generate multi-MB Wasm bundles, compression of which seems pointless and slow), and it would also make sure you compress the post-minification JS bundle.

**Comment by JT117** [comment]
Thanks for the feedback :)
I'm on it, I will look again at rocket to see if I missed something, compress only the js and wasm bundle, change the flag name to
disable_bundle_compression and accept all the corrections you made

File: packages/perseus-rocket/src/pre_compressed_named_file.rs
Comment (JT117) [reply packages/perseus-rocket/src/pre_compressed_named_file.rs:1]: Oh my, I just re looked over, I have a much simpler way, update incoming

**Review by arctic-hen7 — APPROVED** [review]
This is looking really fantastic, thank you so much for all your hard work, it's greatly appreciated!

**Comment by arctic-hen7** [comment]
I've just resolved a tiny conflict (Git being very silly), and, once tests are passing, this should be good to merge! Thanks again!

**Review by arctic-hen7 — APPROVED** [review]
Fabulous, thanks so much for all this!

**Pull request merged**